### PR TITLE
Build everything when outputting everything

### LINKED ISSUE #2092: Outputting a subdirectory should not stop from outputting full package
Our behavior today for `pbr build -o everthing -o web:foo` is broken. We will be missing the compiled `js` output under `everything/test/` for instance. This is because we output `test`, but never built it - `buildDirs` was `web` so we skipped targets under `test` and likely only output source files and non-optional outputs. `pbr serve -o everything web:8080` is likely similarly broken.

Having a `--ouput` that wants to emit the entire package should also mean to build the entire package.

### PR DESCRIPTION
**Pull request by natebosch**
Fixes #2092 
Fixes https://github.com/dart-lang/build/issues/2095

### CODE REVIEW

**Comment by jakemac53** [comment]
I was talking with Gary about this a bit this morning - the existing behavior does allow some things that aren't allowed any more with this change.

Specifically, after this change you can no longer build a single top level dir (`test` for example) without hoisting that directory to the top level. The `test` command currently relies on this behavior.

Previously, `-o` would only cause additional directories to be built if you asked explicitly for it with the `<dir-to-build>:<output-dir>` syntax.

File: build_runner/lib/src/entrypoint/test.dart [hunk 1]
```diff
@@ -108,13 +107,20 @@ class TestCommand extends BuildRunnerCommand {
...
+          .createSync(p.join(testDir.path, 'packages'), recursive: true);
+
+      return await _runTests(tempDir.absolute.uri.toFilePath());
     } on _BuildTestDependencyError catch (e) {
       stdout.writeln(e);
       return ExitCode.config.code;
     } finally {
       // Clean up the output dir.
-      await Directory(tempPath).delete(recursive: true);
+      await Directory(tempDir.absolute.uri.toFilePath())
```
Comment (natebosch) [inline build_runner/lib/src/entrypoint/test.dart]: `await tempDir.delete(recursive: true);`

File: build_runner/lib/src/entrypoint/test.dart [hunk 2]
```diff
@@ -77,15 +77,14 @@ class TestCommand extends BuildRunnerCommand {
   Future<int> run() async {
     SharedOptions options;
     // We always run our tests in a temp dir.
-    var tempPath = Directory.systemTemp
-        .createTempSync('build_runner_test')
-        .absolute
-        .uri
-        .toFilePath();
+    var tempDir = Directory.systemTemp.createTempSync('build_runner_test');
+    var testDir = Directory.fromUri(Uri.parse(p.join(tempDir.path, 'test')));
```
Comment (natebosch) [inline build_runner/lib/src/entrypoint/test.dart]: `var testDirPath = p.join(tempDir.path, 'test');`

Or _maybe_

`var testDirPath = p.join(tempDir.absolute.uri.toFilePath(), 'test');`

File: build_runner_core/lib/src/environment/create_merged_dir.dart [hunk 3]
```diff
@@ -140,7 +140,7 @@ Future<AssetId> _writeAsset(AssetId id, Directory outputDir, String root,
     var outputId = AssetId(packageGraph.root.name, assetPath);
     try {
       if (symlinkOnly) {
-        await Link(_filePathFor(outputDir, id)).create(
+        await Link(_filePathFor(outputDir, outputId)).create(
```
Comment (natebosch) [inline build_runner_core/lib/src/environment/create_merged_dir.dart]: I think this should go out on it's own first.

**Comment by grouma** [comment]
Closing in favor of https://github.com/dart-lang/build/pull/2091

**Pull request closed without merging**